### PR TITLE
Implement constraint registry for transforms and bijections

### LINKED ISSUE #55: Register .transform methods for distribution parameters
PR https://github.com/pytorch/pytorch/pull/4902

Each distribution should know the support of its parameters and results (e.g. `(-inf,inf)`, `(0,inf)`, `(0,1)`). This issue is about registering transform pairs for parameters (unlike distribution outputs, parameters do not need a `.log_abs_det_jacobian()` method and so do not need a full `Bijector` class).

See [relevant section in the Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.unemmumtnzfk). See also #56

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #55 #56
Supersedes #64 
Blocked by pytorch#4771

This implements two lookup tables `transform_to(-)` and `biject_to(-)`. Each inputs a `Constraint` object and returns a `Transform` object that transforms from `constraints.real` to the given constraint. To perform the reverse transform, use `transform_to(c).inv`. This is the final piece of metadata linking all of torch.distributions constraints and transforms.

Here's a [snapshot of the documentation](http://fritzo.org/temp/pytorch/html/distributions.html#module-torch.distributions.constraint_registry).

This design is loosely modeled on constraint-transform registry of Stan. See [Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.unemmumtnzfk) for more details.

- [x] add docs
- [x] add tests

### CODE REVIEW

**Comment by fritzo** [comment]
@alicanb I've added a `ComposeTransform` and an `identity_transform = ComposeTransform([])` in this PR since these are necessary to support the `Transform` interface as the return type of `transform_to()` and `biject_to()`.

**Comment by alicanb** [comment]
Yes, I'm looking at it now, looks beautiful.

**Comment by fritzo** [comment]
@ragulpr @dustinvtran Here is a second draft of a constraint-transform registration mechanism. Design help is always welcome :smile:

File: torch/distributions/transforms.py [hunk 1]
```diff
@@ -155,6 +155,54 @@ def log_abs_det_jacobian(self, x, y):
         return -self.inv.log_abs_det_jacobian(y, x)
 
 
+class ComposeTransform(Transform):
```
Comment (alicanb) [inline torch/distributions/transforms.py]: add `ComposeTransform` to `__all__`?

**Comment by dustinvtran** [comment]
The design spec at the top of `constraint_registry.py` was helpful for understanding. Great job!

Three questions:

1. Can you clarify why we want both`biject_to` and `transform_to`? (Do we dislike polymorphism of accepting either tensor or distribution as input?)
2. Can you describe advantages of a registration mechanism which maintains global state, over simply a pure function with if-else logic for each constraint? E.g., do we expect users to define new constraints and register alternative transforms? Also, which is more Pytorch idiomatic?
3. Unrelated to this PR but the overall design: unconstrained transforms of parameters currently find a given constraint by, e.g., `Normal.params['scale']`. This seems like the wrong containment. Putting my PL hat on, the constraint should be part of a Tensor's type. Are there any long-term plans on how constraints intersect with the rest of Pytorch's codebase?

**Comment by alicanb** [comment]
@fritzo Can you explain the difference between `biject_to` & `transform_to` a little bit (where to use which)?

**Comment by alicanb** [comment]
Probably not in this PR, but we should think about doing `scale = torch.zeros(1).transform_to(constraint)`, it reads much better imo.

File: torch/distributions/constraint_registry.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+@biject_to.register(constraints.interval)
+@transform_to.register(constraints.interval)
+def _transform_to_interval(constraint):
+    span = constraint.upper_bound - constraint.lower_bound
+    loc = -span * constraint.lower_bound
+    scale = span.reciprocal()
+    return transforms.ComposeTransform([transforms.AffineTransform(loc, scale),
+                                        transforms.SigmoidTransform()])
+
+
```
Comment (fritzo) [inline torch/distributions/constraint_registry.py]: @dustinvtran This is the main difference between `transform_to()` and `biject_to()`. If you really need a bijection, then you should use `StickBreakingTransform` for `simplex`. If you don't really need a bijection, then `BoltzmannTransform` is cheaper, more numerically stable, and in theory works better with coordinate-wise optimizers like Adam.

**Comment by fritzo** [comment]
> `torch.zeros(1).transform_to(constraint)` reads much better imo.

That is incorrect for both `.is_cuda` and for GPU placement in multi-GPU systems. Any time you're creating a new tensor, you should do it with `my_existing_tensor.new()` to ensure it lives on the same device.

**Comment by fritzo** [comment]
@dustinvtran Thanks for the questions!

> 1. Can you clarify why we want both `biject_to` and `transform_to`?

`transform_to()` is useful for unconstrained optimization in SVI where you can be really sloppy and overparameterize and thereby get better numerical stability and coordinatewise separation. `biject_to()` is useful for restrictive settings like HMC where you really do need bijections between spaces of the same dimension. The canonical case is `simplex` where `StickBreakingTransform` is needed for HMC but a cheaper sloppier `BoltzmanTransform` (= Exp-then-Normalize) works better for optimization settings. Another case is `LowerCholeskyTransform` where you can be sloppy and `transform_to()` from a matrix with an unused upper triangle, whereas `biject_to()` would require slower reshaping and flattening of the lower triangle to a contiguous vector.

> 2. Can you describe advantages of a registration mechanism which maintains global state, over simply a pure function with if-else logic for each constraint?

Ha ha python code is global state too :smile: But really, here are some advantages I see in a registration mechanism:
- Users (like Pyro and ProbTorch) can define their own `Constraint`s and `Transform`s and register them dynamically with the system. This is the same motivation we have for `kl_divergence()` and `register_kl()`. In Pyro we have lots of custom distributions and a few custom transforms.
- Registration reads more declaratively than imperative if-else logic.
- Table lookup is cheaper than a deep if-else logic.

> Also, which is more Pytorch idiomatic?

From what I've seen, declarative registration seems more PyTorch idiomatic, as e.g. in derivatives.yml. Moreover PyTorch allows extension by hooking into the autograd mechanism with `torch.autograd.Function`. I guess I see Python as so dynamic anyway that this registration mechanism seems relatively static.

> 3. ... the constraint should be part of a Tensor's type. Are there any long-term plans on how constraints intersect with the rest of Pytorch's codebase?

I don't know, maybe @apaszke can comment. While this constraint mechanism is very useful within torch.distributions, it does seem to stray from PyTorch's core abstractions. @apaszke suggested maybe spinning out torch.distributions into a hosted library akin to `torchvision`, and maybe we are getting to a point where this makes sense. I do feel that constraint registration is of wider interest beyond Pyro users (@ragulpr has requested something like this), and I'd like to see it live upstream of Pyro.

**Comment by apaszke** [comment]
Regarding 3. tensors allow attribute assignments, so you could attach constraints to them. I don't think we want to add a "torch.distributions constraint" slot to all tensors, but it shouldn't be necessary since we have attributes.

**Comment by fritzo** [comment]
Ok I've added some more detail to the docs. It would be nice to have some torch.distributions tutorials. We'll definitely be exercising this in Pyro tutorials.

**Comment by fritzo** [comment]
Regarding 3. it looks like this PR is agnostic to the constraint metadata location. I've used `Normal.params['scale']` in the documentation, but as implemented this should be forward compatible with @dustinvtran's idea of annotating tensors with `.constraint` attributes :smile:
```py
unconstrained = transform_to(my_tensor.constraint).inv(my_tensor)
```

File: torch/distributions/transforms.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/transforms.py]: Done.

**Comment by dustinvtran** [comment]
That's awesome. Given that's possible, it would be nice in the future to allow direct application of a tensor or distribution. A high-level API of `transform(tensor)` and `biject(distribution)` seems very natural. Under the hood, it would call `transform_to(my_tensor.constraint)` or `biject_to(my_distribution.support)`.

**Comment by apaszke** [comment]
Good idea! It would be nice to have distributions and tensors both be first class values :+1:

**Comment by fritzo** [comment]
Moved upstream to https://github.com/pytorch/pytorch/pull/4902
